### DISCUSSION TITLE
Using Hairless MIDI

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#23] [Q&A] [answered]
What should be the expected output for Hairless MIDI?

I followed the steps on https://open-e-drums.tumblr.com/post/171304647319/using-hairless-midi and see data coming in but there's a warning about a status byte:

![hairless1](https://user-images.githubusercontent.com/305679/147258639-e3a8091d-1c6f-4879-a3e2-226f47aecd8f.JPG)
![hairless2](https://user-images.githubusercontent.com/305679/147258648-ad4d7c7b-e1cb-4db2-a426-677ed929a4e7.JPG)

I'm using the `thijstriemstra_prototype` branch with the following pads:

| ID | Pad | Type | Triggering | GPIO |
| --- | --- | --- | --- | --- |
| 1 | Drum-tec Diabolo | Snare | Dual | 36, 35 |
| 2 | Roland KD-120BK | Kick | Single | 33 |
| 3 | Roland CY-5 | Hihat | Dual | 32, 26 |
| 4 | Roland FD-8 | Hihat controller | Single | 25 |
| 5 | Roland PD-5 | Tom 1 | Single | 34 |
| 6 | Roland PD-5 | Tom 2 | Single | 39 |

Although only Tom 2 is connected to the ESP32/circuit (is this a problem?)

**Comment by thijstriemstra** [comment, marked as answer]
Turns out I was using the wrong baudrate (should be 38400).

It's now showing a contineous stream of this, and hitting the PD-5 doesn't seem to influence the data (although it's going too fast to properly see):

![hairless3](https://user-images.githubusercontent.com/305679/147259713-8e2ff62e-f6ae-45f3-a4ea-236967b89820.JPG)

**Reply by corrados** [reply]
> Although only Tom 2 is connected to the ESP32/circuit (is this a problem?)

Yes, this is actually a problem for Edrumulus. It expects all configured inputs (i.e. head and rim shot) to have the correct analog circuit connected so that the voltage bias is correctly set.

**Reply by thijstriemstra** [reply]
Thanks for the feedback. I will connect everything and 🤞🏼 

Oh btw, I can safely disconnect the drum pads themselves while testing right, or do I also need to make sure all inputs are connected on that side? Because I do not have all these pads here at the moment..

**Reply by corrados** [reply]
> I can safely disconnect the drum pads themselves while testing right

Yes, you can do that. No problem.

**Reply by thijstriemstra** [reply]
> Thanks for the feedback. I will connect everything and 🤞🏼

Everything's connected and I'm about to plug the USB cable and monitor. Should I merge the master branch into the `thijstriemstra_prototype` branch first @corrados?

**Reply by corrados** [reply]
If you use the ESP32, you do not need to merge from main.

**Reply by thijstriemstra** [reply]
> Everything's connected and I'm about to plug the USB cable and monitor.

After uploading, the serial monitor is silent, where before with incomplete circuit it would spit out data constantly. I connected a PD-5 to TOM1 and for every hit it prints some garbled data in the monitor, which I expected as well. So far so good?

If I use `ttymidi` with this command:

```
ttymidi -s /dev/ttyUSB0 --baudrate=38400 -v
```

The results look much better. The monitor is silent at startup. A single hit on PD-5 on TOM1 prints:

```
Serial	99 31 02	Note on			09 049 002
Serial	89 31 00	Note off		09 049 000
```

Hitting it again:

```
Serial	99 31 01	Note on			09 049 001
Serial	89 31 00	Note off		09 049 000
```

I do see this printed every now and then, although I don't hit it:

```
Serial	99 24 01	Note on			09 036 001
Serial	89 24 00	Note off		09 036 000
```

I will give hairless midi in windows a try next. Where to go from there?

**Reply by corrados** [reply]
Have you installed [Octave](https://www.gnu.org/software/octave/download) to be able to change the Edrumulus pad parameters easily (using edrumuluscontrol.m)?

**Reply by corrados** [reply]
If the "09 049 002" number 49 is the MIDI number, then you have connected the PD-5 to the "crash" input.

**Reply by thijstriemstra** [reply]
> If the "09 049 002" number 49 is the MIDI number, then you have connected the PD-5 to the "crash" input.

Not sure, more about the output of ttymidi on http://www.varal.org/ttymidi/

> Have you installed Octave 

No, will give it a try.

**Reply by corrados** [reply]
Also, if the "002" is the detected velocity, then this value seems to be too low if you actually hit the pad. I expect values much larger like in the range of 60 to 90 if you hit the pad with normal velocity.

**Reply by thijstriemstra** [reply]
I'll only focus on the 3 'easiest' circuits for now:

## Kick using PD-5:
![kick](https://user-images.githubusercontent.com/305679/147707948-6c40952d-9cb5-4415-a45d-249d0eed87ca.jpg)

## Tom1 using PD-5:
![tom1](https://user-images.githubusercontent.com/305679/147707953-e5a96be8-374a-44ee-af14-4e481ca83587.jpg)

## Tom2 using PD-5:
![tom2](https://user-images.githubusercontent.com/305679/147707958-732e2e38-8841-433e-8744-97b4da0d38cd.jpg)

**Reply by thijstriemstra** [reply]
> Also, if the "002" is the detected velocity, then this value seems to be too low if you actually hit the pad. I expect values much larger like in the range of 60 to 90 if you hit the pad with normal velocity.

Sorry, I was tapping it with my finger, and reaching 127 is possible with sticks (just hard on my crappy desk setup atm):


## Tom1 using PD-5:

![velocity](https://user-images.githubusercontent.com/305679/147708344-2eb87311-6dab-4ee3-a103-399d2148296c.jpg)

**Reply by thijstriemstra** [reply]
## Kick using KD120:
![kick-kd120](https://user-images.githubusercontent.com/305679/147708796-452ab8b8-13db-4d70-a98f-0aa6bac53dad.jpg)

**Reply by corrados** [reply]
That looks good :+1:

**Reply by thijstriemstra** [reply]
@corrados it does indeed! I played around with the setup for a few hours last night, and I found everything extremely responsive. There were some 'ghost notes' played here and there (even when I didn't play) but that was to be expected. I guess there are several things I would like to discuss but without any audio/video footage it will be hard to explain. So I'm 3d-printing a proper mount for my iphone now so I can film the drumming. I'm also planning on testing the current prototype with the Teensy 3.6 as it looks like it's becoming usable, yay! To be continued..

**Reply by corrados** [reply]
Sounds very good! It seems to me that you do not regret to have chosen Edrumulus for your project (at least not yet ;-) ).

**Reply by corrados** [reply]
When you are happy with your prototype, it would be great if you could post some information about it in [this thread](https://github.com/corrados/edrumulus/discussions/28) I have just created.

**Reply by thijstriemstra** [reply]
> When you are happy with your prototype, it would be great if you could post some information

Will do!

So I tested more with HairlessMIDI and I find the software acting really strange under load. It's either loading a random heavy program or the USB port that starts triggering HairlessMIDI.

For example, in this video below I have everything connected to a windows machine and as soon as I plug my iphone (white usb cable) in front of the pc, next to the edrumulus serial usb (black usb cable), you can hear as the File explorer for the iphone starts opening, windows loading iphone drivers etc, the hairlessmidi bridge starts going crazy. Uploading a video to github will also drive the midi input insane.

This leads me to believe that I'll never have good results as long as HairlessMIDI is being used (as opposed to either a different controller like teensy or different host software like serialmidi.py). Thoughts?

https://user-images.githubusercontent.com/305679/147892209-d8a88265-64cb-4b4e-b7d5-91e757604988.mp4

**Reply by corrados** [reply]
That might not be caused by HairlessMIDI but by electrical interference from the other MIDI device which interfers with the pad inputs of your ESP32. You could try to increase the velocity threshold and see if you can suppress these false hits by that.

**Comment by thijstriemstra** [comment]
> If the "09 049 002" number 49 is the MIDI number, then you have connected the PD-5 to the "crash" input.

No I think the comments in the code are incorrect, and it correctly hit tom1 aka 49.

I created #27 for this.
